benchmark/algorithms/faiss_t3.py

```diff
@@ -0,0 +1,404 @@
+from __future__ import absolute_import
+import numpy as np
+import sklearn.preprocessing
+import ctypes
+import faiss
+import os
+import time
+import gc
+import resource
+import threading
+import json
+
+from multiprocessing.pool import ThreadPool
+
+from benchmark.algorithms.base import BaseANN
+from benchmark.datasets import DATASETS, download_accelerated
+
+def unwind_index_ivf(index):
+    if isinstance(index, faiss.IndexPreTransform):
+        assert index.chain.size() == 1
+        vt = index.chain.at(0)
+        index_ivf, vt2 = unwind_index_ivf(faiss.downcast_index(index.index))
+        assert vt2 is None
+        return index_ivf, vt
+    if hasattr(faiss, "IndexRefine") and isinstance(index, faiss.IndexRefine):
+        return unwind_index_ivf(faiss.downcast_index(index.base_index))
+    if isinstance(index, faiss.IndexIVF):
+        return index, None
+    else:
+        return None, None
+
+def rate_limited_iter(l):
+    'a thread pre-processes the next element'
+    pool = ThreadPool(1)
+    res = None
+
+    def next_or_None():
+        try:
+            return next(l)
+        except StopIteration:
+            return None
+
+    while True:
+        res_next = pool.apply_async(next_or_None)
+        if res is not None:
+            res = res.get()
+            if res is None:
+                return
+            yield res
+        res = res_next
+
+def build_index(buildthreads, by_residual, maxtrain, clustering_niter, 
+     indexkey, indexfile, add_bs, add_splits, ds, train_on_gpu=True, quantizer_on_gpu_add=True):
```

Comment:
I wonder if T1 and T3 should re-use the same build_index code? It looks almost c&p?

Reply:
I think you are right.  It looks like the only major difference between T1 and T3 is the use of the gpu-quantizer.  Perhaps a new base class could accomplish this.